includes/razorpay-webhook.php

```diff
@@ -116,4 +135,116 @@ protected function paymentAuthorized($data)
 
         exit;
     }
+
+    /**
+     * Currently we handle only subscription failures using this webhook
+     *
+     * @param $data
+     */
+    protected function paymentFailed($data)
+    {
+        //
+        // Order entity should be sent as part of the webhook payload
+        //
+        $orderId = $data['payload']['payment']['entity']['notes']['woocommerce_order_id'];
+
+        $paymentId = $data['payload']['payment']['entity']['id'];
+
+        if (isset($data['payload']['payment']['subscription_id']) === true)
+        {
+            $this->processSubscription($orderId, $paymentId, false);
+        }
+
+        exit;
+    }
+
+    /**
+     * Handling the subscription charged webhook
+     *
+     * @param $data
+     */
+    protected function subscriptionCharged($data)
+    {
+        //
+        // Order entity should be sent as part of the webhook payload
+        //
+        $orderId = $data['payload']['subscription']['entity']['notes']['woocommerce_order_id'];
+
+        $this->processSubscription($orderId);
+
+        exit;
+    }
+
+    /**
+     * Helper method used to handle all subscription processing
+     *
+     * @param $orderId
+     * @param $paymentId
+     * @param $success
+     */
+    protected function processSubscription($orderId, $paymentId, $success = true)
+    {
+        //
+        // If success is false, automatically process subscription failure
+        //
+        if ($success === false)
+        {
+            return $this->processSubscriptionFailed($orderId);
+        }
+
+        // This method is used to process the subscription's recurring payment
+        $wcSubscription = wcs_get_subscriptions_for_order($orderId);
+
+        $subscriptionId = get_post_meta($orderId, self::RAZORPAY_SUBSCRIPTION_ID)[0];
+
+        $api = $this->razorpay->getRazorpayApiInstance();
+
+        $subscription = $api->subscription->fetch($subscriptionId);
+
+        $this->processSubscriptionSuccess($wcSubscription, $subscription, $paymentId);
```

Comment:
How about:

`$this->processSubscriptionSuccess($wcSubscription[0], $subscription[0], $paymentId);`

The method signature asks for `"$subscription"`, which doesn't feel like an array object at all. Much cleaner to pass a required single object rather than an array of exactly one object.